cli/src/pcluster/templates/cdk_builder.py

```diff
@@ -12,15 +12,151 @@
+        archecture = "x86_64"
+
+        # Setup ImageBuilder Resources
+        # InstanceRole
+        # TODO: use imagebuild.config.build_config.instance_role to create custom one
+        instancerole = iam.CfnRole(
+            self,
+            id="InstanceRole",
+            managed_policy_arns=[
+                core.Fn.sub("arn:${AWS::Partition}:iam::aws:policy/AmazonSSMManagedInstanceCore"),
+                core.Fn.sub("arn:${AWS::Partition}:iam::aws:policy/EC2InstanceProfileForImageBuilder"),
+            ],
+            assume_role_policy_document={
+                "Statement": {
+                    "Action": "sts:AssumeRole",
+                    "Effect": "Allow",
+                    "Principal": {"Service": "ec2.amazonaws.com"},
+                },
+                "Version": "2012-10-17",
+            },
+            path="/executionServiceEC2Role/",
+        )
+        instancerole.add_metadata("Comment", "Role to be used by instance during image build.")
+
+        # InstanceProfile
+        iam.CfnInstanceProfile(
+            self, id="InstanceProfile", path="/executionServiceEC2Role/", roles=[core.Fn.ref("InstanceRole")]
+        )
+
+        # TODO add S3 bucket
+
+        # InfrastructureConfiguration
+        imagebuilder.CfnInfrastructureConfiguration(
+            self,
+            id="PClusterImageInfrastructureConfiguration",
+            name="PCluster-Image-Infrastructure-Configuration",
+            instance_profile_name=core.Fn.ref("InstanceProfile"),
+            terminate_instance_on_failure=imagebuild.config.build_config.terminate_on_failure,
+            instance_types=["p2.xlarge", "c5.xlarge"] if archecture == "x86_64" else ["m6g.xlarge"],
+            key_pair=imagebuild.config.build_config.key_pair_name,
+        )
+
+        # Component
+        if imagebuild.config.dev_settings_config and imagebuild.config.dev_settings_config.update_and_reboot:
+            imagebuilder.CfnComponent(
+                self,
+                id="PClusterComponent",
+                name="PCluster",
+                version="0.0.1",
+                description="Bake PCluster AMI",
+                change_description="First version",
+                platform="Linux",
+                # TODO: Fix the bug using data property or uri with core.Fn.sub
+                uri="s3://test-slurm/official_ami_build_component.yml",
+            )
+        else:
+            imagebuilder.CfnComponent(
+                self,
+                id="PClusterComponent",
+                name="PCluster",
+                version="0.0.1",
+                description="Bake PCluster custom AMI",
+                change_description="First version",
+                platform="Linux",
+                # TODO: Fix the bug using data property or uri with core.Fn.sub
+                uri="s3://test-slurm/custom_ami_build_component.yml",
+            )
+
+        # ImageRecipe
+        imagebuilder.CfnImageRecipe(
```

Comment:
You could think to define a `ImageRecipeConstruct` to isolate the code.

Reply:
I don't fully understand the Construct, the ImageRecipe relies on components defined above, so should I move both components and ImageRecipe to  `ImageRecipeConstruct`? I notice HeadNode is an object attribute of Cluster, so we create a construct for it. In this case, should we create an object like Recipe for ImageBuilder?